### PR TITLE
Improve job consistency [HZ-2753]

### LINKED ISSUE #19257: getStatus() for job reference of finished light job should throw JobNotFoundException
When light job is finished then `getStatus()` method should throw `JobNotFoundException`.

Reproducer:
```
HazelcastInstance inst = createHazelcastInstance();
HazelcastInstance client = createHazelcastClient();
Job job1 = inst.getJet().newLightJob(streamingDag());
assertJobExecuting(job1, inst);
Job job2 = client.getJet().getJob(job1.getId());
job1.cancel();
assertThatThrownBy(job1::join)
        .isInstanceOf(CancellationException.class);

sleepSeconds(1);
System.out.println(job2.getStatus());
assertThatThrownBy(job2::getStatus)
        .isInstanceOf(JobNotFoundException.class);
```

### PR DESCRIPTION
**Pull request by burakgok**
This PR eliminates inconsistent job proxy behaviors, which are originated from the lazy initialization of job completion future, by joining and using the future whenever possible (db30e2c). Joining is performed in a synchronous manner by sending a second operation, which is expected to throw `JobNotFoundException` if the job is completed and removed. This operation could simply be `GetJobIdsOperation`, but since `JetGetJobIdsMessageTask` is multi-target, it wouldn't be efficient and that's why `GetJobStatusOperation` is extended for light jobs.

This change, however, does not resolve inconsistencies stemmed from reordered operations, which are exemplified by:
```java
DAG dag = new DAG().vertex(new Vertex("test", () -> new MockP().streaming()));
Job job = client().getJet().newLightJob(dag);
cancelAndJoin(job);
```
In the above case, the submission and termination requests are sometimes reordered on the coordinator since
1. client requests return as soon as they are sent (unlike member requests), and
2. requests can be handled by different coordinator threads.

The ordering problem is alleviated [by differentiating _uninitialized_ and _removed_ jobs][1], but the race-free solution requires assigning each master context to a fixed coordinator thread (9e2a005, #23082). To implement the solution, `StripedExecutor` is converted into a `ManagedExecutorService` and `submitToCoordinatorThread` is made to submit `StripedCallable`s. However, since we don't have access to `Runnable`s produced internally by `CompletionStage` methods (`thenCompose()`, `handleAsync()`, etc.), `StripedRunnable`s are replaced by _worker-views_ of `StripedExecutor` (bfa9aa4). Finally, each worker is converted to a limited `ExecutorService`, which improves the performance a little bit (2de4220). 

**Notes:**
1. To facilitate review process, _formatting_ and _refactoring_ changes are isolated from behavioral changes.
2. Unneeded locks are not yet removed.

Fixes #19257
EE PR: hazelcast/hazelcast-enterprise#6287
Protocol changes: hazelcast/hazelcast-client-protocol#469

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases

[1]: https://github.com/hazelcast/hazelcast/blob/2de42209b5f55bf6cfd5d7beda404500a7e962af/hazelcast/src/main/java/com/hazelcast/jet/impl/JobCoordinationService.java#L379-L381

### CODE REVIEW

File: hazelcast/src/test/java/com/hazelcast/test/HazelcastTestSupport.java [hunk 1]
```diff
@@ -365,6 +349,38 @@ public static void interruptCurrentThread(final int delayMillis) {
...
+    /**
+     * Computes the cartesian product of the values of the specified maps and
+     * prepends each result array with the comma-separated list of corresponding keys.
+     */
+    @SafeVarargs
+    public static List<Object[]> parameters(Map<String, ?>... maps) {
+        return Sets.cartesianProduct(
+                        Arrays.stream(maps).map(Map::entrySet).collect(toList())
+                ).stream().map(pairs ->
+                        Stream.concat(
```
Comment (TomaszGaweda) [inline hazelcast/src/test/java/com/hazelcast/test/HazelcastTestSupport.java:377]:
```suggestion
new Object[2] { 
// same  joining not wrapped in Stream.of,
Util.toList(pairs, Entry::getValue)
}
```

toList() can be pulled from Jet's Util class (which can be not-only-Jet now)

File: hazelcast/src/test/java/com/hazelcast/jet/core/JobCompletionFutureTest.java [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+            List<String> names = new ArrayList<>(applicableThrowers);
+            names.add(0, names.remove(i));
+
+            testTrackedJob((job, trackedJob) -> {
+                NoOutputSourceP.proceedLatch.countDown();
+                job.join();
+                for (int j = 0; j < names.size(); j++) {
+                    String name = names.get(j);
+                    assertThatThrownBy(() -> methods.get(name).accept(trackedJob),
+                                    String.join(" -> ", names.subList(0, j + 1)))
```
Comment (TomaszGaweda) [inline hazelcast/src/test/java/com/hazelcast/jet/core/JobCompletionFutureTest.java:162]: Please extract String.join to a variable, will be easier to read

File: hazelcast/src/test/java/com/hazelcast/jet/core/JobCompletionFutureTest.java [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+            assertTrueAllTheTime(() -> assertThatCode(() -> trackedFuture.get().get()).doesNotThrowAnyException(), 2);
+            assertCompletion(trackedJob);
+        });
+    }
+
+    @Test
+    public void test_waitForCompletion() throws InterruptedException {
+        for (String name : triggers) {
+            assumeFalse(useLightJob && asList("suspend()", "resume()", "restart()").contains(name));
+            System.out.println(">> Will call " + name + " to join future and then wait for completion");
```
Comment (TomaszGaweda) [inline hazelcast/src/test/java/com/hazelcast/jet/core/JobCompletionFutureTest.java:192]: logger? :)

File: hazelcast/src/test/java/com/hazelcast/jet/core/JobCompletionFutureTest.java [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+        semaphores.put(key, semaphore);
+        return semaphore;
+    }
+
+    private static void waitIfBlocked(String key) {
+        Semaphore semaphore = semaphores.remove(key);
+        if (semaphore != null) {
+            try {
+                semaphore.acquire();
+            } catch (InterruptedException ignored) { }
```
Comment (TomaszGaweda) [inline hazelcast/src/test/java/com/hazelcast/jet/core/JobCompletionFutureTest.java:295]: catching InterruptedException means you should interrupt current thread

File: hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/listener/ClientListenerServiceImpl.java [hunk 5]
```diff
@@ -153,12 +153,12 @@ public void provideStaticMetrics(MetricsRegistry registry) {
...
     @Probe(name = CLIENT_METRIC_LISTENER_SERVICE_EVENT_QUEUE_SIZE, level = MANDATORY)
     private int eventQueueSize() {
-        return eventExecutor.getWorkQueueSize();
+        return eventExecutor.getQueueSize();
     }
 
     @Probe(name = CLIENT_METRIC_LISTENER_SERVICE_EVENTS_PROCESSED, level = MANDATORY)
     private long eventsProcessed() {
-        return eventExecutor.processedCount();
+        return eventExecutor.getCompletedTaskCount();
```
Comment (Fly-Style) [inline hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/listener/ClientListenerServiceImpl.java:161]: nit: I suggest you to not use "get" prefix, if field doesn't have setter. Just `eventExecutor.completedTaskCount()`

File: hazelcast/src/main/java/com/hazelcast/internal/util/executor/CompletableFutureTask.java [hunk 6]
```diff
@@ -35,12 +36,12 @@ public class CompletableFutureTask<V> extends DeserializingCompletableFuture<V>
 
     private volatile Thread runner;
 
-    CompletableFutureTask(Callable<V> callable) {
+    CompletableFutureTask(@Nonnull Callable<V> callable) {
```
Comment (Fly-Style) [inline hazelcast/src/main/java/com/hazelcast/internal/util/executor/CompletableFutureTask.java:39]: Not all callers guarantee the null-safety. E.g,  `CachedExecutorServiceDelegate#sumbit`

File: hazelcast/src/main/java/com/hazelcast/internal/util/executor/CompletableFutureTask.java [hunk 7]
```diff
@@ -57,16 +58,9 @@ public void run() {
...
-            if (c != null) {
-                Object result = null;
-                try {
-                    result = c.call();
-                    complete((V) result);
-                } catch (Throwable ex) {
-                    completeExceptionally(ex);
-                }
-            }
+            complete(callable.call());
```
Comment (Fly-Style) [inline hazelcast/src/main/java/com/hazelcast/internal/util/executor/CompletableFutureTask.java:61]: What do you think about preserving original code? I am sure, it's easier to do a null-check here, than guarantee null-safety for all callers.

File: hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/listener/ClientListenerServiceImpl.java (hunk 5 above)
Comment (Fly-Style) [reply hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/listener/ClientListenerServiceImpl.java:161]: I see it comes from `ManagedExecutorService`, closing my answer.

File: hazelcast/src/main/java/com/hazelcast/internal/util/executor/StripedExecutor.java [hunk 8]
```diff
@@ -222,19 +316,17 @@ private long timeoutNanos(Runnable task) {
         @Override
         public void run() {
             try {
-                while (live) {
+                while (!(shutdown.get() && Thread.interrupted())) {
```
Comment (Fly-Style) [inline hazelcast/src/main/java/com/hazelcast/internal/util/executor/StripedExecutor.java:319]: Is it thread-safe?

File: hazelcast/src/test/java/com/hazelcast/test/HazelcastTestSupport.java [hunk 9]
```diff
@@ -929,6 +951,22 @@ public static void assertJoinable(long timeoutSeconds, Thread... threads) {
         }
     }
 
+    protected static <T> void assertCollection(Collection<T> expected, Collection<T> actual) {
```
Comment (k-jamroz) [inline hazelcast/src/test/java/com/hazelcast/test/HazelcastTestSupport.java:954]: AssertJ has similar assertions, we should use them

**Pull request closed without merging**